docs/developers/architecture/app_model.md

```diff
@@ -265,11 +421,113 @@ be used easily as a command in a menu, or bound to a keybinding.  It is up to
 napari to determine what providers it will make available, and what type hints
 plugins/users may use to request dependencies.
 
+### Annotation and namespaces
+
+For parameter annotations to be resolved by in-n-out, the object must be in the
+"global" or "local" namespace. The "global" namespace consists of the items
+in the callback functions {py:attr}`funciton.__globals__` attribute.
+In in-n-out the "local" namespace consists of objects in the {py:mod}`typing`
+and {py:mod}`types` modules' namespace, and objects in the
+napari {attr}`~app_model.Application.injection_store`'s `namespace` attribute.
+In the napari `app`s `Store`, some basic napari objects are added to `namespace`
+attribute:
+
+* all public types from {mod}`napari.components`, {mod}`napari.layers`
+  and {mod}`napari.viewer`
+* {class}`~napari._qt.qt_main_window.Window`
+* {class}`~napari._qt.qt_viewer.QtViewer`
+
+This means that when annotating an callback function, you will not need to import
+any of the above classes for the annotation to be resolved. For other classes
+you will need to import them at the **module** level. Importing in any of the following
+ways will not work:
+
+* import within `typing.TYPE_CHECKING`
+* import within an outer function, where the callback function is an inner function
+
+### Providers and processors in napari
+
+Non-Qt providers and processors are defined in
+[`napari/_app_model/injection`](https://github.com/napari/napari/tree/main/napari/_app_model/injection).
+Qt providers and processors are defined in
+[`napari/_qt/_qapp_model/injection`](https://github.com/napari/napari/tree/main/napari/_qt/_qapp_model/injection).
+
+Non-Qt providers and processors are registered in the `app` `Store` during
+initialization of the napari `app`, in `NapariApplication`'s
+{meth}`~napari._app_model._app.NapariApplication.__init__`.
+Qt providers and processors are registered in
+{func}`~napari._qt._qapp_model.qactions.init_qactions`, which gets called during
+initialization of `_QtMainWindow`. This is the same as
+[registration of `Action`s](app-model-actions-napari).
+
+## app-model testing
+
+This section provides a guide to testing app-model aspects of napari. For general
+information on napari testing see [](napari-testing).
+
+### Mock app
+
+The {class}`~napari._app_model._app.NapariApplication`, `app`, is a global application
+singleton. This is not ideal because changes made in a previous test will persist and
+can cause conflicts for subsequent tests, when running a suit of tests.
+Segmentation faults can also occur as a
+singleton `app` may keep a reference to an object, e.g., a
+{class}`~napari._qt.qt_main_window.Window`, that has
+since been cleaned up at the end of a previous test.
+Thus we mock the `app` and
+[autouse](https://docs.pytest.org/en/7.1.x/how-to/fixtures.html#autouse-fixtures-fixtures-you-don-t-have-to-request)
+it so a new instance of `app` is returned everytime {func}`napari._app_model.get_app`
+is used inside a test.
```

Comment:
```suggestion
is used inside a test. Note that a new `app` instance is also created when the `make_napari_viewer` fixture is used.
```

Reply:
`get_app` is called in `init_qactions` which is called in `make_napari_viewer` so year you're right. It is not returned though.

Reply:
So it seems that `_mock_app` is called during setup of any test (meaning a `app` is created for every test?). Is this because of the `autouse` ?

Reply:
(yes it is due to `autouse` that `_mock_app` fixture is called during setup of every test)

Reply:
I've added a note about this.